쥬스 메이커 [Step 1] 우롱차, vayne

dnwhd0112 opened the pull request:
@AppleCEO

## 코드를 작성할때 고민되었던 점
1. FruitStore를 Class로 사용한 이유? 
JuiceMaker에서 프로퍼티로 FruitStore 를 가지는대  고유한 FruitStore 인스턴스를 가져야 하기 때문이다. class는 할당할때 call by reference로 할당 하기 때문
참고 문서: https://developer.apple.com/documentation/swift/choosing_between_structures_and_classes
의 Use Classes When You Need to Control Identity

2. JuiceMaker의 recipe를 생성자로부터 모든걸 할당할지 아님 default 값을 줄지 고민하였습니다.
전자의 경우 인스턴스를 만들때 마다 필요한것 인스턴스들을 전부 다 생성해야되는 불편함이 있습니다. 반면 후자는 custom 하기가 어렵다는 단점이 있습니다. 저희의 경우 Common 쪽에 defaultJuiceRecipe를 생성하여 이를 할당하는 방향으로 설계하였습니다. 

## 조언을 얻고 싶은 부분

네이밍 부분과 좀 더 좋은 설계에 대하여 조언을 얻고 싶습니다.

sky891 committed:
feat: Fruit, Juice, JuiceMakingError 열거형 추가

앞으로 사용할 열거형들의 타입 정의

sky891 committed:
docs: Common 파일 추가

sky891 committed:
feat: 과일 수량 체크 함수 추가

isEnoughStock 함수

sky891 committed:
feat: 과일 소모 함수 추가

useFruit 함수

sky891 committed:
feat: 과일 추가 함수 추가

addFruit 함수, useFruit 함수 로직 수정

sky891 committed:
feat: 새로운 과일 추가 함수, 과일 삭제 함수 추가

addNewFruit, removeFruit 함수

dnwhd0112 committed:
feat: 에러 케이스 추가

noRecipe 케이스 추가

dnwhd0112 committed:
refactor: 네이밍 수정 및 반환 타입 수정

isEnoughStock -> checkStock, return type bool -> void

dnwhd0112 committed:
feat: JuiceMaker 프로퍼티 추가, 기능 추가

recipe 프로퍼티, fruitStore, 생성자 추가, make 함수 ,getAJuiceRecipe 함수 추가

dnwhd0112 committed:
Create README.md

sky891 committed:
refactor: common 파일 수정

네이밍 변경

dnwhd0112 committed:
refactor: FruitStore 리펙토링

네이밍 변경

dnwhd0112 committed:
refactor: JuiceMaker 리펙토링

Common 파일에 defaultJuiceRecipe 추가 및 ErrorMessage 열거형 추가
JuiceMaker  생성자 변경 및 네이밍 변경, 에러 메세지 출력시 열거형 사용하도록 변경.

AppleCEO commented on JuiceMaker/JuiceMaker.xcodeproj/project.pbxproj:
```diff
@@ -171,6 +174,7 @@
 				C73DAF3B255D0CDD00020D38 /* ViewController.swift in Sources */,
 				C73DAF37255D0CDD00020D38 /* AppDelegate.swift in Sources */,
 				C73DAF39255D0CDD00020D38 /* SceneDelegate.swift in Sources */,
+				EA30E04527C4BD79003F668B /* Common.swift in Sources */,
```
Common 이라는 파일명은 조금 위험해보이네요.
저희 회사에도 Common 이라는 파일이 있는데요.
어디에서 쓰이는 코드든 모아놓고 있어서 수천줄이 되서 골치에요.
명확한 파일명을 가지고 여러 개로 나누는 것이 더 좋을 것 같습니다~

dnwhd0112 replied:
그럼 enum 별로 Fruit, Juice, Error 파일을 나뉘는게 좋을꺼 같네요!

AppleCEO commented on JuiceMaker/JuiceMaker/Model/Common.swift:
```diff
@@ -0,0 +1,40 @@
+
+enum Fruit {
+    case strawberry, banana, pineapple, kiwi, mango
+}
```
과일의 종류를 enum 으로 모아놓았군요!
이렇게 여러 공통된 것들을 모아놓는데 enum은 훌륭한 대안이 됩니다~ 👍

AppleCEO commented on JuiceMaker/JuiceMaker/Model/Common.swift:
```diff
@@ -0,0 +1,40 @@
+
+enum Juice {
+    case strawberryTaste, bananaTaste, kiwiTaste, pineappleTaste, strawberryBannaTaste, mangoTaste, mangoKiwiTaste
+    static let defaultJuiceRecipe = [Juice.strawberryTaste:[Fruit.strawberry: 16],
```
주스 레시피들을 모아놓았군요! 상수 이름에 default를 붙인 이유는 무엇인가요?

dnwhd0112 replied:
JuiceMaker 별로 다른 레서피를 가질 수 있다고 생각해서 쥬스들의 기본이 되는 레서피를 만들어 주었습니다.

AppleCEO commented on JuiceMaker/JuiceMaker/Model/Common.swift:
```diff
@@ -0,0 +1,40 @@
+                                     Juice.mangoKiwiTaste:[Fruit.mango: 2, Fruit.kiwi: 1]]
+}
+
+enum JuiceMakingError: Error {
+    case notEnoughStock
+    case notRegisteredFruit
+    case duplicatedFruit
+    case noRecipe
+}
+
+enum ErrorMessage {
+    static let notEnoughStock = "재고가 부족합니다."
+    static let notRegisteredFruit = "존재하지 않는 과일입니다."
+    static let noRecipe = "레서피가 존재하지 않습니다."
+    static let unkown = "알 수 없는 오류가 발생했습니다."
+}
```
에러들을 모아놓은 부분도 좋습니다~ 🔢 
에러 메시지들을 따로 열거형으로 만들어보셨네요!
이 에러 메시지들을 JuiceMakingError 에 옮겨보는 것은 어떨까요?

dnwhd0112 replied:
그 방법도 좋은거 같네요! 다만 변수명은 좀 수정을 해야겠어요!

AppleCEO commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -8,5 +8,58 @@ import Foundation
 
 // 과일 저장소 타입
 class FruitStore {
+    private let name: String
```
과일 저장소가 여러개 될 수 있는건가요? ㅎㅎ

dnwhd0112 replied:
그 부분이 명확하지가 않아서 여러개의 FruitStore가 존재 할 수 있다고 생각하고 설계했습니다.

AppleCEO replied:
너무 확장성이 높은 것도 좋은 것만은 아닙니다!
스텝이 이렇게 나눠져있는 것도 우선 작은 기능부터 제대로 만들도록 의도한 거라고 생각합니다~
다음 스텝을 생각하지 않고 현재 스텝의 요구사항만 충족하는 간단한 프로그램부터 만들어보는 것 어떨까요?

AppleCEO replied:
(우선 하나의 FruitStore를 제대로 만들면 이거를 여러개 만드는 것은 어렵지 않을 것입니다)

dnwhd0112 replied:
알겠습니다 일단 name 부분은 뺄게요!

AppleCEO commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -8,5 +8,43 @@ import Foundation
 
 // 쥬스 메이커 타입
 struct JuiceMaker {
+    private var recipe: Dictionary<Juice,Dictionary<Fruit,Int>> = Juice.defaultJuiceRecipe
```
레시피 상수를 만든 이유가 있을까요? 필요할 때 Juice.defaultJuiceRecipe 에서 바로 가져올 수도 있지 않을까요?

dnwhd0112 replied:
JuiceMaker의 인스턴스 마다 다르게 레서피를 가질 수 있다고 생각해서 프로퍼티에 레서피 상수를 할당했습니다.(step 1 만 있어서 추후에 어떻게 필요한지 정확히 몰라서요 ㅠ). 사실 레서피 변경 메소드 또한 추가하고 싶었는대 요구사항에는 존재하지 않아서 추가하지는 않았어요.

AppleCEO commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -8,5 +8,43 @@ import Foundation
+    private var recipe: Dictionary<Juice,Dictionary<Fruit,Int>> = Juice.defaultJuiceRecipe
+    private let fruitStore: FruitStore
     
+    init(fruitStore: FruitStore) {
+        self.fruitStore = fruitStore
+    }
```
밖에서 과일 저장소를 할당 받을 수 있군요! 💯 
의존성 주입을 의도한 걸까요?

dnwhd0112 replied:
네! 맞습니다.

AppleCEO commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -8,5 +8,43 @@ import Foundation
+        self.fruitStore = fruitStore
+    }
+    
+    func getAJuiceRecipe(juice: Juice) throws -> Dictionary<Fruit,Int> {
+        guard let juiceRecipe = self.recipe[juice] else {
+            throw JuiceMakingError.noRecipe
+        }
+        return juiceRecipe
+    }
```
실패할 수 있는 코드에 throws 를 사용해 보셨군요~ 🥇

AppleCEO commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -8,5 +8,43 @@ import Foundation
+        return juiceRecipe
+    }
+    
+    func makeJuice(_ juice: Juice) -> Juice? {
+        do {
+            let juiceRecipe = try getAJuiceRecipe(juice: juice)
+            for (fruit,amount) in juiceRecipe {
+                try self.fruitStore.checkStock(fruit: fruit, amount: amount)
+            }
+            for (fruit,amount) in juiceRecipe {
+                try self.fruitStore.consumeStock(fruit: fruit, amount: amount)
+            }
+            return juice
+        } catch JuiceMakingError.notEnoughStock {
+            print(ErrorMessage.notEnoughStock)
+            return nil
+        } catch JuiceMakingError.notRegisteredFruit {
+            print(ErrorMessage.notRegisteredFruit)
+            return nil
+        } catch JuiceMakingError.noRecipe {
+            print(ErrorMessage.noRecipe)
+            return nil
+        }
+        catch {
+            print(ErrorMessage.unkown)
+            return nil
+        }
+    }
```
의미없는 공백이 있지 않아서 좋네요~ 👍 
반면에 코드가 길고 조금 복잡해보이네요. 
제 생각에 에러를 받는 부분을 하나로 각 에러마다 에러 미시지를 프린트해주게 수정해줄 수 있을 것 같은데요.
조금 더 나아가서는 에러 메시지를 출력하는 메소드를 만들어서 그 메소드에서 에러를 출력해볼 수도 있겠어요~

dnwhd0112 replied:
오.... 시도해본적 없는 방법이네요! 구글링해서 찾아보고 구현해보겠습니다!

AppleCEO reviewed (changes requested):
안녕하세요. 도미닉입니다.
@dnwhd0112, @vayne77 고생하셨습니다~
코드들이 간결하고 대부분 보기 좋았습니다!
몇가지 구조적인 질문들 코멘트 남겨봤습니다.
확인해주시고 개선해보세요.

dnwhd0112 committed:
refactor: 1차 피드백 개선

Common 파일 -> Fruit, Juice, JuiceMakingError 으로 나눔
errorMessage를 JuiceMakingError 내부에 존재하도록 변경
errorMessage 출력함수 생성
FruitStore에 name 프로퍼티 삭제

AppleCEO commented on JuiceMaker/JuiceMaker/Model/Juice.swift:
```diff
@@ -1,16 +1,12 @@
 //  JuiceMaker
 //
-//  Created by 김영훈 on 2022/02/22.
+//  Created by Vayne, 우롱차 on 2022/02/24.
```
작성자를 바꾼 부분 좋습니다~

AppleCEO commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,25 +1,18 @@
 //
 //  JuiceMaker - FruitStore.swift
-//  Created by yagom. 
+//  Created by Vayne, 우롱차. 
 //  Copyright © yagom academy. All rights reserved.
```
작성자를 바꾼 부분 좋습니다! 이런 작은 부분도 컨트롤할 수 있으면 더 좋은 개발자가 될 수 있을 겁니다~

AppleCEO commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,25 +1,18 @@
     private var fruitStock: Dictionary<Fruit, Int>
     
-    init(name: String, fruitStock: Dictionary<Fruit, Int>) {
-        self.name = name
```
네임을 제거해보셨군요!
확장성을 생각해도 좋지만 그것이 너무 과하면 오히려 프로그램이 복잡해질 수 있습니다.
아직은 하나의 과일 저장소, 하나의 과일 재고, 하나의 쥬스 메이커로도 충분한 시점이라고 생각하네요.
기본적인 것이 훌륭하게 되면 앞으로 여러 개로 바꾸는 것은 금방 필요할 때 늘리고 줄이고 할 수 있을 것입니다~

dnwhd0112 replied:
처음부터 너무 많은걸 생각한거 같아요... ㅋㅋㅋ

AppleCEO commented on JuiceMaker/JuiceMaker/Model/Juice.swift:
```diff
@@ -21,20 +17,3 @@ enum Juice {
-    static let notRegisteredFruit = "존재하지 않는 과일입니다."
-    static let noRecipe = "레서피가 존재하지 않습니다."
-    static let unkown = "알 수 없는 오류가 발생했습니다."
-}
```
열거형을 따로 빼내보셨군요! 💯 
하나의 열거형, 구조체, 클래스마다 하나의 파일에 저장하는 것이 좋습니다~

AppleCEO commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -32,18 +31,8 @@ struct JuiceMaker {
-            return nil
-        }
-        catch {
-            print(ErrorMessage.unkown)
+        } catch {
+            JuiceMakingError.printErrorMessage(error: error as? JuiceMakingError ?? JuiceMakingError.unkownError)
             return nil
         }
```
좋네요! 
코드가 깔끔해졌습니다. 🥇 
다만 물음표가 많아서 조금 복잡해보이네요.
여기서 ?? 로 기본값을 주지 않고 옵셔널 바인딩을 해보면 어떨까요?

dnwhd0112 replied:
네!

AppleCEO commented on JuiceMaker/JuiceMaker/Model/JuiceMakingError.swift:
```diff
@@ -0,0 +1,34 @@
+    static let unkownErrorMessage = "알 수 없는 오류가 발생했습니다."
+    
+    static func printErrorMessage(error: JuiceMakingError) {
+        if error == JuiceMakingError.notEnoughStock {
+            print(JuiceMakingError.notEnoughStockErrorMessage)
+        } else if error == JuiceMakingError.notRegisteredFruit {
+            print(JuiceMakingError.notRegisteredFruitErrorMessage)
+            
+        } else if error == JuiceMakingError.noRecipe {
+            print(JuiceMakingError.noRecipeErrorMessage)
+        } else {
+            print(JuiceMakingError.unkownErrorMessage)
+        }
```
이 반복문이 이 쪽으로 옮겨왔군요. 👍 
이렇게 하는 것도 개선이 되어서 좋지만 switch 문 등을 사용해서 케이스마다 에러 메시지를 출력해주면 어떨까요?

dnwhd0112 replied:
아무래도 기존 코드를 재사용 하다 보니까 이렇게 됬네요 ㅋㅋ... 제 생각에도 switch 문이 더 나은거 같아요. 수정하도록 하겠습니다.

AppleCEO reviewed (changes requested):
안녕하세요 도미닉입니다.
@dnwhd0112, @vayne77 고생하셨습니다~
몇가지 코멘트 남겨봤습니다.
이번 스텝에서 진행할지, 아니면 다음 스탭에서 반영할지는 이야기 나눠보아요!

dnwhd0112 committed:
refactor: 피드백 2차 반영

쥬스 메이킹 에러 메세지 출력함수 -> switch 문으로 변경
Error 타입 변환시 if let 구문으로 변경
